### PR TITLE
Delay Git exception

### PR DESCRIPTION
**Pull request by Radinyn**

### CODE REVIEW

File: protostar/git.py [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+
+    @staticmethod
+    def clone(path_to_repo: Path, repo_to_clone: "GitRepository"):
+        path_to_repo.parent.mkdir(parents=True, exist_ok=True)
+        cloned_repo = GitRepository(path_to_repo)
+        cloned_repo.clone(repo_to_clone.path_to_repo)
+        return cloned_repo
+
+    @staticmethod
+    def from_existing(path_to_repo: Path):
```
Comment (kasperski95) [inline protostar/git.py:69]: - `from_*` naming convention is used by static factories. It suggests creating `Git` not `GitRepository`. Maybe `create_from_initialized`
- `path_to_repo` could be provided through the constructor

File: protostar/git.py [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+        "stderr": subprocess.DEVNULL,
+    }
+)
+
+CREDENTIALS = [
+    "-c",
+    'user.name="Protostar"',
+    "-c",
+    'user.email="protostar@protostar.protostar"',
+]
```
Comment (kasperski95) [inline protostar/git.py:30]: I think we should use the user credentials.
(Next.js uses user credentials for the first commit.)

File: protostar/git.py [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+                "add",
+                str(path_to_item),
+            ],
+            **SHARED_KWARGS,
+            cwd=self.path_to_repo,
+        )
+
+    # pylint: disable=invalid-name
+    @wrap_git_exception
+    def rm(self, path_to_item: Path, force: bool = False):
```
Comment (kasperski95) [inline protostar/git.py:170]:
```suggestion
    def remove_submodule(self, submodule_name: str):
```
Just to be consistent with, `get`, `add`, and `update` submodule and to narrow down the use case.

File: protostar/git.py [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+    'user.email="protostar@protostar.protostar"',
+]
+
+
+class ProtostarGitException(ProtostarException):
+    def __init__(self, message: str, details: Optional[str] = None):
+        super().__init__("Error while executing Git command:\n" + message, details)
+
+
+class NotARepositoryException(ProtostarGitException):
```
Comment (kasperski95) [inline protostar/git.py:38]: We can skip articles. I'd rename this exception to:

```suggestion
class InvalidGitRepositoryException(ProtostarGitException):
```

File: protostar/git.py [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+    def wrapper(*args, **kwargs):
+        try:
+            func(*args, **kwargs)
+        except subprocess.CalledProcessError as ex:
+            raise ProtostarGitException(str(ex)) from ex
+
+    return wrapper
+
+
+class Git:
```
Comment (kasperski95) [inline protostar/git.py:53]: It could be moved to a separate file.

File: protostar/git.py (hunk 4 above)
Comment (kasperski95) [inline protostar/git.py:38]: Exceptions could be at the bottom of a file.


Vertical formatting: https://levelup.gitconnected.com/javascript-clean-code-vertical-formatting-d15097f5f30e

File: protostar/git.py [hunk 6]
```diff
@@ -0,0 +1,215 @@
+import subprocess
```
Comment (kasperski95) [inline protostar/git.py:1]: This class could be tested.

File: protostar/git.py [hunk 7]
```diff
@@ -0,0 +1,215 @@
...
+    def from_existing(path_to_repo: Path):
+        path = GitRepository.get_repo_root(path_to_repo)
+        return GitRepository(path)
+
+
+class GitRepository:
+    def __init__(self, path_to_repo: Path):
+        """
+        This class should not be instantiated directly, use `Git.init/clone/from_existing` instead.
+        """
```
Comment (kasperski95) [inline protostar/git.py:78]: Is this comment necessary?

**Review by kasperski95 — APPROVED** [review]

**Comment by kasperski95** [comment]
You can resolve threads in another PR.

File: protostar/git.py (hunk 2 above)
Comment (Radinyn) [reply protostar/git.py:30]: I'm not sure whether this will work because we pipe stdout to dev/null

**Pull request merged**